### DISCUSSION TITLE
Module admin_settings page not loading

### DISCUSSION THREAD

**Discussion by eawmaas** [xenocrat/chyrp-lite#198] [Q&A] [answered]
I am still exploring Chyrp and I am building a module called headadditions  that gives admins the opportunity to add scripts like Google Analytics in the head-section of the page.

I use the develop-branch with DEBUG on (no errors in the log)

The module does work already, I include a file from the root (Chyrp is in a subdirectory) which contains all the scripts that I want to include and the scripts do load).

I managed to get a tab in admin-settings, I want to show a textfield where admins can paste scripts or a php-include which then will be loaded in the headsection.

But when I click on the headadditions tab in admin I get a blank page with
Not Found
The requested resource was not found.

I am puzzling with this all day now and I have tried everything I think, cache is off.

The module has three files:

headadditions/info.php:

```
<?php
return array(
    "name"          => __("Head Additions", "headadditions"),
    "url"           => "http://chyrplite.net/",
    "version"       => "2023.09",
    "description"   => __("Makes it possible to add items at the end of the documents head-section.", "head_additions"),
    "author"        => array(
        "name"      => "Erwin Maas",
        "url"       => "https://exdomo.com/"
    )
);
```

headadditions/headadditions.php:

```
<?php
    class HeadAdditions extends Modules {

        public function headadditions_settings($admin): void {
            if (!Visitor::current()->group->can("change_settings"))
                show_403(
                    __("Access Denied"),
                    __("You do not have sufficient privileges to change settings.")
                );
    
            if (empty($_POST)) {
                $admin->display("pages".DIR."headadditions_settings");
                return;
            }
    
            if (!isset($_POST['hash']) or !authenticate($_POST['hash']))
                show_403(
                    __("Access Denied"),
                    __("Invalid authentication token.")
                );
            }

       public function settings_nav($navs): array {
            if (Visitor::current()->group->can("change_settings"))
                $navs["headadditions_settings"] = array(
                    "title" => __("Headadditions", "headadditions_settings")
                );

            return $navs;
        }

        public function end_head() {
		include "../myownincludes/headercodes.php";
        }
		
    }
```

and headadditions/admin/pages/headadditions_settings.twig:

```
{% extends "layouts" ~ DIR ~ "default.twig" %}

{% block title %}{{ "Headadditions Settings" | translate("headadditions") }}{% endblock %}

{% block content %}
<h2>{{ "Headadditions Settings" | translate("headadditions") }}</h2>
<form id="headadditions_settings" class="split" action="{{ url('headadditions') }}" method="post" accept-charset="UTF-8">

<fieldset role="presentation">

<p>
<label for="headadditions_settings">{{ "Enter item(s) or include a file containing items" | translate("headadditions_settings") }}
</label>
<input type="text" name="headadditions_settings" id="headadditions_settings"{{ site.module_headadditions }}>
</p>

<p class="buttons">
<button type="submit" class="yay"><img src="{{ site.chyrp_url }}/admin/images/icons/success.svg" alt="">{{ "Update" | translate }}</button>
</p>
<input type="hidden" name="hash" value="{{ authenticate() }}" id="hash">
</fieldset>
</form>
{% endblock %}
```

I feel I am so close to get the form on the admin-settingspage but why does it throw the Not-Found error?
What have I missed?

**Comment by xenocrat** [comment, marked as answer]
Rename your function `headadditions_settings()` to `admin_headadditions_settings()`. You can read about why that is necessary on [this wiki page](https://github.com/xenocrat/chyrp-lite/wiki/Routes-and-Controllers).

**Comment by eawmaas** [comment]
Thank you! This should help me go further!
I have read the Wiki several times but could not make that connection.

**Comment by xenocrat** [comment]
Happy to help. I link to the relevant wiki page just to help you join the dots. ☺️

**Reply by eawmaas** [reply]
Yes that's excellent, just a hint so I can explore further and learn. Doing it yourself is the best way and I don't want to use hacky solutions, I already have my scripts in the head but that was the easy part, 30 minutes or so.
Now I need to understand the "Chyrp-way" ;-) and I did not see that the route-part was where my problems were and I am already making progressions again :thumb:
